### DISCUSSION TITLE
Ordered dithering with custom palette

### DISCUSSION THREAD

**Discussion by page200** [ImageMagick/ImageMagick#2370] [Help] [answered]
Can ImageMagick perform ordered dithering using specified colors?

In case it helps, here's what I found so far:

 - `MagickOrderedDitherImage` has no option for custom colors.

 - `MagickRemapImage` has dither methods to choose from, but none of them is ordered dithering.

 - All of the following commands yield different results:

   1. `magick convert input.png -ordered-dither o8x8,7 output1.gif`
   1. `magick convert input.png -ordered-dither o8x8,7 output.gif`
   1. `magick convert input.png -ordered-dither o8x8 -remap palette.gif output.gif`
   1. `magick convert input.png -ordered-dither o8x8,7 -remap palette.gif output.gif`
   1. `magick convert input.png -remap palette.gif -ordered-dither o8x8 output.gif`
   1. `magick convert input.png -remap palette.gif -ordered-dither o8x8,7 output.gif`

   It seems that none of these commands use colors from `palette.gif` to perform ordered dithering. (Because on one hand, `,7` in command 4 means that 7 specific color levels should be used, on the other hand `-remap palette.gif` means that colors from `palette.gif` should be used. So I would imagine that one of these two contradictory requests would be ignored, and the result would be the same as for command 2 or 3, but it's different from both.)
   So it seems that commands 3 and 4 first perform dithering with standard colors and then remap using colors from `palette.gif`, whereas commands 5 and 6 first remap and then dither.

I couldn't find a way to perform ordered dithering **using** custom colors.

**Comment by snibgo** [comment, marked as answer]
> Can ImageMagick perform ordered dithering using specified colors?

I suppose you mean "colours specified in the command (or API)". Sadly, no.

IM performs operations in the order give on the command. "-ordered-dither" and "-remap" are operations. Whichever comes first in the command will be executed first. They are not settings for each other. The second operation works on the output of the first, but otherwise doesn't "know" that the previous operation has occurred.

> ... So I would imagine that one of these two contradictory requests would be ignored, ...

No. Both requests are performed.

**Reply by JohnChristianD** [reply]
Are two resize operators also completely separate functions so something like
-distort Resize 200% -define -filter Lanczos -distort Resize 1366x768
 or 
-define -filter Lanczos -distort Resize 200% -distort Resize 1366x768
 , the first and last respectively would use the default Robidoux filter, correct? Or is there something I'm missing with the syntax here?

**Comment by snibgo** [comment]
@JohnChristianD: Your syntax is incorrect: remove the text "-define" from your commands.

"-filter Lanczos" is a setting that will be used by any "-distort Resize" operations that occur *after* the filter setting. Any resizes that occur *before* the filter setting will use the default.

**Reply by JohnChristianD** [reply]
Does the -define filter:lobes=4 work for every -resize or -distort Resize Lanczos after that or only one? Also additional question, do equally sharp much larger resolution images produce sharper downscales ceteris paribus if outputting a non-percentage resolution i.e. x720?
